include/foxy/detail/tunnel.hpp

```diff
@@ -97,51 +108,207 @@ struct tunnel_op : boost::asio::coroutine
+                                : s.uri_parts.scheme();
+
+          auto port = s.uri_parts.port().size() == 0 ? static_cast<std::string>(scheme)
+                                                     : static_cast<std::string>(s.uri_parts.port());
+
+          s.client.async_connect(static_cast<std::string>(s.uri_parts.host()), std::move(port),
+                                 bind_handler(std::move(*this), on_connect_t{}, _1, _2));
+        }
+
+        if (ec) {
+          s.response->result(http::status::bad_request);
+          s.response->body() =
+            "Unable to connect to the remote at: " + static_cast<std::string>(s.uri_parts.host()) +
+            "\nError code: " + ec.message() + "\n\n";
+
+          s.response->prepare_payload();
+
+          BOOST_ASIO_CORO_YIELD
+          s.server.async_write(*s.response, std::move(*this));
+
+          if (ec) { goto upcall; }
+          if (s.parser->get().keep_alive()) { continue; }
+          break;
+        }
+
+      } else {
+        std::cout << "sending malformed response message\n";
+
+        s.response->result(http::status::bad_request);
+        s.response->body() =
+          "Malformed client request. Use either CONNECT <authority-uri> or <verb> <absolute-uri>";
+        s.response->prepare_payload();
+
+        BOOST_ASIO_CORO_YIELD
+        s.server.async_write(*s.response, std::move(*this));
+
+        if (ec) { goto upcall; }
+        if (s.parser.get().keep_alive()) { continue; }
+        break;
+      }
+
+      if (s.is_absolute && s.is_http) {
+        s.parser->get().keep_alive(false);
+
+        BOOST_ASIO_CORO_YIELD
+        {
+          auto target =
+            s.uri_parts.path().size() == 0
+              ? (s.parser->get().method() == http::verb::options ? boost::string_view("*")
+                                                                 : boost::string_view("/"))
+              : s.uri_parts.path();
+
+          s.parser->get().target(target);
+
+          async_relay(s.server, s.client, std::move(*s.parser),
+                      bind_handler(std::move(*this), on_relay_t{}, _1, _2));
+        }
+
+        if (ec) { goto upcall; }
+        s.close_tunnel = true;
+        break;
+      }
+
+      std::cout << "going to write back the tunnel response now\n";
+
+      s.response->result(http::status::ok);
+
+      BOOST_ASIO_CORO_YIELD
+      s.server.async_write(*s.response, std::move(*this));
+
+      if (ec) {
+        s.close_tunnel = true;
+        goto upcall;
+      }
+
+      s.close_tunnel = false;
+      break;
+    }
+
+    {
+      auto const guard        = std::move(s.work);
+      auto const close_tunnel = s.close_tunnel;
+      return p_.invoke(boost::system::error_code(), close_tunnel);
+    }
+
   upcall:
     if (!is_continuation) {
       BOOST_ASIO_CORO_YIELD
-      boost::asio::post(bind_handler(std::move(*this), ec, 0));
+      net::post(bind_handler(std::move(*this), ec, 0));
```

Comment:
1-arg post might not be the right thing to do here @djarek ?

Reply:
1-arg post is correct in this context, because `tunnel_op` uses the intrusive executor hooks.